titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeApiServerIntegrator.java

```diff
@@ -424,4 +550,160 @@ private void publishContainerEvent(String taskId, TaskState taskState, String re
+        List<V1Pod> orphanedPodsToGc = currentPods.stream()
+                .filter(p -> {
+                    String nodeName = p.getSpec().getNodeName();
+                    return StringExt.isNotEmpty(nodeName) && !currentNodesNames.contains(nodeName);
+                })
+                .collect(Collectors.toList());
+
+        // GC orphaned pods on nodes that are no longer available
+        logger.debug("Attempting to GC {} orphaned pods without valid nodes", orphanedPodsToGc.size());
+        for (V1Pod pod : orphanedPodsToGc) {
+            gcPod(pod);
+        }
+        logger.debug("Finished orphaned pod GC without valid nodes");
+
+        Set<String> currentPodNames = currentPods.stream().map(p -> p.getMetadata().getName()).collect(Collectors.toSet());
+        Set<String> currentTaskIds = currentTasks.keySet();
+        List<V1Pod> podsStuckBeingDeleted = currentPods.stream()
+                .filter(pod -> {
+                    String podName = pod.getMetadata().getName();
+                    DateTime deletionTimestamp = pod.getMetadata().getDeletionTimestamp();
+                    return clock.isPast(deletionTimestamp.getMillis() + POD_GC_DELETION_TTL) &&
+                            !currentTaskIds.contains(podName);
+                })
+                .collect(Collectors.toList());
+
+        // GC orphaned pods that stuck being deleted
+        logger.debug("Attempting to GC {} orphaned pods not in Titus", podsStuckBeingDeleted.size());
+        for (V1Pod pod : podsStuckBeingDeleted) {
+            gcPod(pod);
+        }
+        logger.debug("Finished orphaned pod GC not in Titus");
+
+
+        List<Task> tasksNotInApiServer = currentTasks.values().stream()
+                .filter(t -> TaskState.isRunning(t.getStatus().getState()) && !currentPodNames.contains(t.getId()))
+                .collect(Collectors.toList());
+
+        // Transition orphaned tasks to Finished that don't exist in Kubernetes
+        logger.debug("Attempting to transition {} orphaned tasks", tasksNotInApiServer.size());
+        for (Task task : tasksNotInApiServer) {
+            publishContainerEvent(task.getId(), Finished, REASON_TASK_LOST, "Inconsistency between control plane and machine", Optional.empty());
+        }
+        logger.debug("Finished orphaned tasks transitions");
+    }
+
+    private void gcNode(V1Node node) {
+        String nodeName = node.getMetadata().getName();
+        long startTimeMs = clock.wallTime();
+        try {
+            normalApi.deleteNode(nodeName, null, null, null, 0, null, "Background");
+            nodesClientMetrics.incrementOnSuccess(DELETE, NODES, STATUS_200);
+        } catch (JsonSyntaxException e) {
+            // this is probably successful. the generated client has the wrong response type
+            nodesClientMetrics.incrementOnSuccess(DELETE, NODES, STATUS_200);
```

Comment:
Is there anything to include in logs or metrics to determine if it "probably" was successful or not?

Reply:
This client usage is temporary until I write the `WebClient` version which will have the correct behavior.